test/fiber/test_mutex.rb

```diff
@@ -194,7 +194,7 @@ def test_queue_pop_waits
   end
 
   def test_mutex_deadlock
-    error_pattern = /No live threads left. Deadlock\?/
+    error_pattern = /lock already owned by another fiber/
```

Comment:
FWIW the `No live threads left. Deadlock?` error is actually an eFatal IIRC.
But I think using ThreadError as you did seems fine too.

There is some other detection too using that message though, maybe it's good to be consistent with it in the exception class?

Reply:
Probably we also need to respect `GET_THREAD()->vm->thread_ignore_deadlock` (which defaults to false) like in `rb_check_deadlock()`, i.e., the lock could be interrupted by a signal and unblock things.

Reply:
> FWIW the `No live threads left. Deadlock?` error is actually an eFatal IIRC.

Interesting, the other deadlock related errors in this function are all `rb_eThreadError`. 

> Probably we also need to respect GET_THREAD()->vm->thread_ignore_deadlock

Done.

Reply:
> Interesting, the other deadlock related errors in this function are all `rb_eThreadError`.

Right, ThreadError seems fair then.